src/lib/lossless_cast.hpp

```diff
@@ -0,0 +1,229 @@
+std::enable_if_t<std::is_same_v<int64_t, Source> && std::is_same_v<int32_t, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  if (source < std::numeric_limits<int32_t>::min() || source > std::numeric_limits<int32_t>::max()) {
+    return std::nullopt;
+  } else {
+    return static_cast<Target>(source);
+  }
+}
+
+// Int to Long
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<int32_t, Source> && std::is_same_v<int64_t, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return static_cast<Target>(source);
+}
+
+// NULL to anything but NULL
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<NullValue, Source> && !std::is_same_v<NullValue, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return std::nullopt;
+}
+
+// Anything but NULL to NULL
+template <typename Target, typename Source>
+std::enable_if_t<!std::is_same_v<NullValue, Source> && std::is_same_v<NullValue, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return std::nullopt;
+}
+
+// String to Integral
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<pmr_string, Source> && std::is_integral_v<Target>, std::optional<Target>> lossless_cast(
+    const Source& source) {
+  static_assert(std::is_same_v<int32_t, Target> || std::is_same_v<int64_t, Target>, "Expected int32_t or int64_t");
+
+  // We don't want std::stol's exceptions to occur when debugging, thus we're going the c-way: strtol, which
+  // communicates conversion errors via errno. See http://c-faq.com/misc/errno.html on why we're setting `errno = 0`
+  errno = 0;
+  char* end;
+
+  const auto integral = std::strtol(source.c_str(), &end, 10);
+
+  if (errno == 0 && end == source.data() + source.size()) {
+    return lossless_cast<Target>(integral);
+  } else {
+    return std::nullopt;
+  }
+}
+
+// String to Floating Point
+// NOT SUPPORTED: Some strings (e.g., "5.5") have lossless float representations, others (e.g., "5.3") do not. Allowing
+//                String to Float conversion just sets up confusion why one string was convertible and another was not.
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<pmr_string, Source> && std::is_floating_point_v<Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return std::nullopt;
+}
+
+// Integral to String
+template <typename Target, typename Source>
+std::enable_if_t<std::is_integral_v<Source> && std::is_same_v<pmr_string, Target>, std::optional<Target>> lossless_cast(
+    const Source& source) {
+  return pmr_string{std::to_string(source)};
+}
+
+// Floating Point to String
+template <typename Target, typename Source>
+std::enable_if_t<std::is_floating_point_v<Source> && std::is_same_v<pmr_string, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  // TODO(moritz) find a lossless float-to-string converter
```

Comment:
Should be impossible given that not all floats can be expressed as a terminating decimal

Reply:
I know you are probably right, but could you give me a quick example?

Reply:
https://stackoverflow.com/questions/1089018/why-cant-decimal-numbers-be-represented-exactly-in-binary

Reply:
Or try entering .1 here: https://www.h-schmidt.net/FloatConverter/

Reply:
this function is about FLOAT to STRING, not the other way round! Of course I know that 0.1 has no exact representation in binary...

Reply:
...but is there any binary float that has no exact representation in decimal. E.g., is there an `n` for which `2^(-n)` is no representable in decimal?

Reply:
2^(-n) always ends with 5 in decimal, so 2^(-n-1) should, too.

Reply:
How would you express the float representation of .1 as a string?

Reply:
Ah, guess I'm wrong. Still, I am not sure if we want to start staring into that abyss: https://stackoverflow.com/questions/7153979/algorithm-to-convert-an-ieee-754-double-to-a-string

Reply:
>  I am not sure if we want to start staring into that abyss: 

Absolutely not sure, either. Which is why I left it open for the moment. I would just declare string-to-float as too cumbersome to do losslessly and thus leave this specialization returning std::nullopt.

Reply:
I decided we do not want to stare into that abyss, lossless float-string conversion is not supported.